Fix #159: Introduce question player UI structure

BenHenning opened the pull request:
Fix #159.

This introduces full support for the question player, minus early exit support (which will be a follow-up work item post-prototype).

This PR is primarily refactoring the state fragment presenter into a shared StatePlayerRecyclerViewAssembler object that can be used to arrange the recycler views for both explorations and questions. This is done in a compositional way that facilitates future features being added, or disparity in functionality between the two (e.g. no previous button in questions and no replay button in explorations).

This PR also includes reworking how the GCS resource bucket name is provided to be normalized (we were using two different annotations before), and set up such that we can change the bucket used depending on context (questions require the test server bucket rather than the production one).

#503 tracks adding thorough tests for the question player once #89 is resolved. This wasn't done as part of this PR for a few reasons:
1. Although #495 shows that we can make reasonable tests pass for these async players, they aren't reliable and have issues (e.g. no Robolectric support)
2. This PR is already quite large and complicated
3. #495 is introducing basic testing for the state player, much of which is shared with questions

This replaces #411 by building on the work started by @jamesxu0.

Open issues that should be resolved after this PR is done:
- [ ] End of question player has weird scrolling (to be finished as part of #40)
- [ ] Not following mocks (to be finished as part of #40)

vinitamurthi committed:
Initial check in for the question data controller

vinitamurthi committed:
Removed progress controller test

BenHenning committed:
Address the renames suggested in #217.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-interface

BenHenning committed:
Introduce thoroughly stubbed QuestionAssessmentProgressController +
empty test suite.

vinitamurthi committed:
Review changes

vinitamurthi committed:
Review changes

vinitamurthi committed:
Pass data providers to the assessment controller, and set caps in the training controller

vinitamurthi committed:
Fixed merge conflicts

vinitamurthi committed:
Removed unnecessary imports/variables

vinitamurthi committed:
Add fake question data for stubbed interfaces

vinitamurthi committed:
Remove duplicate questions while fetching in training controller

vinitamurthi committed:
Comment explaining the filter function

vinitamurthi committed:
Improve duplicate checking - check it while filtering instead of after filtering

vinitamurthi committed:
Add linked skill id values

vinitamurthi committed:
Review changes

vinitamurthi committed:
add a new module for questions constants

vinitamurthi committed:
fix merge conflicts

vinitamurthi committed:
Review changes

vinitamurthi committed:
add a test to verify questions were fetched properly

vinitamurthi committed:
reformatted code

BenHenning committed:
Merge branch 'question_data' into introduce-question-progress-controller-interface

Conflicts:
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt
	model/src/main/proto/question.proto

BenHenning committed:
Re-add QuestionTrainingController removed in merge.

BenHenning committed:
Finalize question progress controller interface.

BenHenning committed:
Implement QuestionAssessmentProgressController.

This includes some basic refactoring of internal structures used by the
exploration progress controller to share common functionality between the
two progress controllers. There's still some duplication, but this seems
like a reasonable split since there's likely to be further differences in
the progress controllers in the future.

The question assessment progress controller tests pass, but no new ones
have yet been added to thoroughly test the implementation.

BenHenning committed:
Add initial phase of tests for the progress controller based on
ExplorationProgressController. They haven't yet been verified as correct.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-interface

Conflicts:
	domain/src/main/assets/sample_questions.json
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/main/java/org/oppia/domain/util/JsonAssetRetriever.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Fix typo in QuestionTrainingController.

BenHenning committed:
Post-merge fixes and adjustments.

BenHenning committed:
Fix broken ProfileManagementControllerTest.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-interface' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Merge branch 'fix-broken-profile-management-controller-test' into introduce-question-progress-controller-implementation

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Update tests to build, but not pass.

jamesxu0 committed:
Merge branch 'introduce-question-progress-controller-interface' into question_player_lofi

jamesxu0 committed:
Started UI

jamesxu0 committed:
setup ViewModel

jamesxu0 committed:
added binding adapter

jamesxu0 committed:
Handles basic functionality

jamesxu0 committed:
Moved code around

jamesxu0 committed:
Added replay button

jamesxu0 committed:
Merge branch 'develop' into question_player_lofi

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/main/java/org/oppia/domain/util/JsonAssetRetriever.kt
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Merge branch 'develop' into question_player_lofi

Conflicts:
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/res/layout/question_player_fragment.xml
	app/src/sharedTest/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivityTest.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Make questions partially work.

This introduces back button behavior to ensure training sessions can be
reset.

This also attempts to introduce proper asset loading, but it doesn't
seem to be working currently.

BenHenning committed:
Temporarily remove check in StateDeck that breaks questions.

BenHenning committed:
Ensure all state navigation buttons are on the same line and the same
size.

This required reworking how nav button binding works in the state
fragment such that multiple models are used depending on which type of
navigation scenario the fragment is currently in.

BenHenning committed:
Fix dp adjustment that caused previous button to disappear.

Fix previous button showing up twice for continue interaction.

BenHenning committed:
Undo .idea/misc.xml change that snuck in.

BenHenning committed:
Update & fix StateFragmentTest (at least for Espresso). See #495 description
for details on the changes necessary to make these tests work.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	app/src/main/res/values/colors.xml

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Post-merge fixes with some temporary stop-gap solutions.

BenHenning committed:
Fix questions image rendering and consolidate GCS resource bucket names.

BenHenning committed:
Fix question session not properly being reinitialized for new sessions.

Also, ensure that notifications do not regenerate the training session.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Fix index accounting for questions. Before, the index incremented upon
correct answer submission rather than navigation.

BenHenning committed:
Remove support for backward navigation in questions since it isn't
allowed.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/state/StateDeck.kt

BenHenning committed:
Set title of question activity to 'Train Mode'.

BenHenning committed:
Add support for a synthetic ephemeral question to represent the end of a
training session.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	model/src/main/proto/question.proto

BenHenning committed:
Update question progress tracking to be translatable and to ensure that
the synthetic terminal state shows the correct progress: 'completed'.

BenHenning committed:
Move most state player functional support into a new configurable
assembler class to enable sharing between the exploration and question
players.

BenHenning committed:
Hook up StatePlayerRecyclerViewAssembler to the question player fragment
presenter. This also enables the congratulations text.

BenHenning committed:
Introduce proper terminal page and disable overscroll. Note that the end
session page introduced here has not been mocked yet, so it's mostly a
placeholder until a mock is available.

BenHenning committed:
Add support & button for training session replay.

BenHenning committed:
Update/add TODOs to correspond to GitHub tracked issues.

BenHenning committed:
Fix broken import.

BenHenning commented:
NB: this iteration doesn't yet match https://xd.adobe.com/view/6384f092-e16a-4214-5a1a-f10ee0ef4f04-15b5/screen/2b1cbe89-66b9-4e2b-bbe3-28bcbd0c293c/TM-Q6-Text-Input-w-Units-Feedback-/. We need to inline the content & left align the answer box for it to more closely match the mocks.

It might be worth following this up in #40 instead of this PR, especially since this PR is quite large as it is.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SubmittedAnswerViewModel.kt:
```diff
@@ -3,4 +3,6 @@ package org.oppia.app.player.state.itemviewmodel
-class SubmittedAnswerViewModel(val submittedUserAnswer: UserAnswer): StateItemViewModel(ViewType.SUBMITTED_ANSWER)
+class SubmittedAnswerViewModel(
+  val submittedUserAnswer: UserAnswer, val gcsEntityId: String
+): StateItemViewModel(ViewType.SUBMITTED_ANSWER)
```
Nit: spacing issue: `) : StateItemViewModel`

BenHenning replied:
About 1/3 of classes currently use this spacing, so keeping it this way. I suggest we just use the linters as the enforcement here once they're broadly enabled.

rt4914 commented on app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt:
```diff
@@ -19,6 +26,27 @@ class QuestionPlayerActivity : InjectableAppCompatActivity() {
+  }
+
   companion object {
     // TODO(#159): Use this skillList from TopicTrainFragment to fetch questions and start train mode.
```
I think this TODO can be removed now.

nikitamarysolomanpvt replied:
+1

BenHenning replied:
Nice catch. Done.

rt4914 commented on app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+    return recyclerViewAssembler.getPendingAnswerHandler(itemList)?.getPendingAnswer()
+      ?: UserAnswer.getDefaultInstance()
+  }
+}
```
EOF missing

BenHenning replied:
Done.

rt4914 commented on app/src/main/res/drawable/progress_bar.xml:
```diff
@@ -0,0 +1,18 @@
+      </shape>
+    </scale>
+  </item>
+</layer-list>
```
EOF missing

nikitamarysolomanpvt replied:
+1

BenHenning replied:
I believe this has been resolved.

rt4914 reviewed (approved):
The code LGTM. 

@BenHenning I have a doubt related to product. In our earlier discussion it came up that the Question Player is not a conversation. So will this implementation be updated in future? If not, then the exploration and question player might become confusing.

nikitamarysolomanpvt reviewed (approved):
LGTM

veena14cs commented on app/src/main/java/org/oppia/app/topic/overview/TopicOverviewFragmentPresenter.kt:
```diff
@@ -26,12 +27,13 @@ class TopicOverviewFragmentPresenter @Inject constructor(
   private lateinit var topicId: String
   private val htmlParser: HtmlParser by lazy {
-    htmlParserFactory.create(/* entityType= */"topic", topicId, /* imageCenterAlign= */ true)
+    htmlParserFactory.create(resourceBucketName, /* entityType= */"topic", topicId, /* imageCenterAlign= */ true)
```
nit change `/* entityType= */ "topic"` spacing.

BenHenning replied:
Nice catch. Done.

veena14cs reviewed (approved):
LGTM

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

BenHenning committed:
Finish tests for Question{Training,AssessmentProgress}Controllers.

Note that the nature of the assessment progress controller is such that
a new, dynamic DataProvider is needed since the data provider for the
current assessment question is actually a transformation of different
data providers depending on which questions are passed to the
controller.

BenHenning committed:
Initial introduction of test coroutine dispatchers to replace Kotlin's
test coroutine dispatcher.

This includes introducing a test-only module to contain testing
dependencies.

BenHenning committed:
Introduce a new LiveData mechanism to bridge coroutines from
DataProviders in a way that doesn't have the same limitations as the
previous MutableLiveData-based bridge.

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	domain/src/test/java/org/oppia/domain/audio/CellularAudioDialogControllerTest.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-question-progress-controller-implementation

BenHenning committed:
Introduce new nested data provider that effectively allows a
DataProvider to be set up like transformAsync() but with the ability to
change the root DataProvider.

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

nikitamarysolomanpvt committed:
question player progress-bar hifi

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
revert

nikitamarysolomanpvt committed:
revert

nikitamarysolomanpvt committed:
Merge remote-tracking branch 'upstream/introduce-question-player-ui-structure' into question-player-progress-bar-hifi

BenHenning committed:
Early work at introducing FakeSystemClock tests (not yet complete).

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	utility/src/main/java/org/oppia/util/data/DataProviders.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

Conflicts:
	app/src/main/AndroidManifest.xml
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt
	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/answerhandling/InteractionAnswerHandler.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelFactory.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelModule.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivityPresenter.kt
	app/src/main/java/org/oppia/app/testing/ContentCardTestActivity.kt
	app/src/main/java/org/oppia/app/testing/ContentCardTestActivityPresenter.kt
	app/src/main/res/layout/my_downloads_activity.xml
	app/src/main/res/layout/state_button_item.xml
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/app/topic/overview/TopicOverviewFragmentTest.kt

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateViewModel.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelFactory.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelModule.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
	app/src/main/java/org/oppia/app/player/stopplaying/StopExplorationDialogFragment.kt
	app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt
	app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragment.kt
	app/src/main/res/values/colors.xml
	app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt
	domain/src/main/java/org/oppia/domain/topic/TopicListController.kt
	domain/src/test/java/org/oppia/domain/topic/TopicListControllerTest.kt
	utility/src/main/java/org/oppia/util/parser/HtmlParser.kt

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

Conflicts:
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/res/layout/state_button_item.xml
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix the submit button disabled states: with the new refactor, the submit
button is now properly disabled when an invalid answer is pending.

BenHenning committed:
Revert accidental changes to .idea/misc.xml.

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	data/src/test/java/org/oppia/data/persistence/PersistentCacheStoreTest.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Remove unnecessary meta file for Mockito.

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Minor import cleanup.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into introduce-question-player-ui-structure

This merge resolution includes a simplification/re-write of hint &
solution handling for explorations (and in a way that might be
easily transferable to questions).

Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateViewModel.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix audio playback & generalize it to facilitate enabling it for
questions in the future.

BenHenning committed:
Address reviewer comments & fix broken tests.

BenHenning committed:
Fix hints & solutions functionality after simplification (some bugs &
regressions were introduced due to the simplification; these have been
fixed).

BenHenning committed:
Revert the assembler delays to match the intended durations rather than
the test-only values used for development.

BenHenning commented:
I updated this to develop and made the following additional changes:
- I rewrote the hints & solutions set up part that was in the presenter to hopefully be a bit simpler, but more so that it can be shared with questions by adding it as another feature to the new state recycler view assembler (some additional work is needed before this will work fully). In the process, I also fixed the following bugs that I found:
  1. Hints would get stuck with a 'new hint' notification when rotating the device without resolving an existing hint.
  2. The automatic "show hint after 30 seconds" would show hints starting at index 2 instead of 1.
- I hooked up the disable-submit-button-on-invalid-input logic to questions to have parity with the exploration player (so this is additional functionality added to the new state recycler view assembler class).
- I moved some of the audio playback coordination logic to the new state recycler view assembler to fix it after merging with develop, but also to facilitate audio playback for questions (similar to hints & solutions, some additional work needs to be done to finish this such as integrating the audio bar & selection the correct GCS entity ID from which to load the voiceovers).

This is ready for review again. PTAL.

veena14cs commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -13,7 +13,7 @@ import org.oppia.app.viewmodel.ObservableArrayList
-  val explorationId: String,
+  val entityId: String,
   interaction: Interaction,
   private val interactionAnswerReceiver: InteractionAnswerReceiver,
```
`interactionAnswerReceiver ` is not been used anywhere.

veena14cs commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -13,7 +13,7 @@ import org.oppia.app.viewmodel.ObservableArrayList
+  val entityId: String,
   interaction: Interaction,
   private val interactionAnswerReceiver: InteractionAnswerReceiver,
   interactionAnswerErrorReceiver: InteractionAnswerErrorReceiver
```
`interactionAnswerErrorReceiver ` is not been used anywhere.

BenHenning replied:
Nice catch, removed.

veena14cs commented on app/src/main/java/org/oppia/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -0,0 +1,886 @@
+      // it will be reset for each subsequent answer.
+      val nextUnrevealedHintIndex = getNextHintIndexToReveal(state)
+      val wrongAnswerCount = pendingState.wrongAnswerList.size
+      if (wrongAnswerCount == 0) {
```
If the user doesn't submit any answer after revealing 1 or 2 hints. It doesn't show up any new hints.

BenHenning replied:
Sorry, is that the behavior you observe when running this code or is that what the behavior should be?

My understanding from reading the original implementation was that we only trigger hints automatically after 30 seconds if the user doesn't do anything after 60 seconds once one wrong answer is submitted, and after 10 seconds if multiple wrong answers are submitted. Note that these timers reset if you actually show a hint since that triggers a change to EphemeralState, so if you don't answer and see the first hint, another will show up after 30 seconds if no activity. I intended to implement the functionality to match this.

Is my understanding incorrect? If so, what part of the above isn't correct?

veena14cs replied:
I observed it while running the code. I am confused here, how it has to behave.When the user doesn't interact or submit any wrong answer after revealing 1 hint. Whether it has to show next hint in 30secs or there shouldn't be any kind of response..

BenHenning replied:
What is the current behavior? From testing it a few weeks ago it seemed like we showed another hint after no activity & showing the first hint, but it's not clear if that's correct? Do you know who on Oppia web we can ask about what the expected behavior should be? Was there someone you worked with to define the behavior currently implemented in develop?

veena14cs replied:
We can ask Kevin from Oppia web. In earlier implementation it was like if user doesn't submit any wrong answer next hint would appear in 30secs.

BenHenning replied:
Thanks Veena. Yeah, I think this implementation is slightly wrong. I'm going to take another stab at it, and see if I can represent the diagram a bit more exactly in code to make this clearer.

veena14cs replied:
Ack :)

BenHenning replied:
Okay, done. It took me a while last week to finish this because I wanted to make sure we thoroughly tested all the different interesting states that can come up. The actual implementation turned out to be pretty quick & simple, and I'm more confident it's correct at this point. #1273 is tracking adding tests, and these will be added in a follow-up PR that I'm sending out shortly.

veena14cs replied:
This works perfectly now. Thanks Ben!

veena14cs commented on app/src/sharedTest/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivityTest.kt:
```diff
@@ -27,7 +27,7 @@ class QuestionPlayerActivityTest {
   @Test
   fun testQuestionPlayerActivity_loadQuestionPlayerFragment_hasDummyString() {
```
Can we remove this testcase, if we aren't showing any dummy text.

BenHenning replied:
Good point. Replaced it with a temp test pointing to #503 to fill in actual functionality. I kept the test suite, though, as a placeholder for the future tests to go.

veena14cs replied:
Sounds reasonable.

veena14cs commented on domain/src/test/java/org/oppia/domain/topic/TopicListControllerTest.kt:
```diff
@@ -38,17 +38,17 @@ import org.oppia.app.model.ProfileId
 import org.oppia.util.threading.BlockingDispatcher
 import org.robolectric.annotation.Config
-import java.util.Date
+import java.util.*
```
Remove wildcard import.

nikitamarysolomanpvt replied:
+1

BenHenning replied:
Done. It's unfortunate that Android Studio sometimes defaults to wildcards despite being configured to never use them. :/

BenHenning replied:
Per time constraints, marking this as done. Please let me know if you disagree.

veena14cs reviewed (changes requested):
Suggested some small changes.

nikitamarysolomanpvt reviewed (approved):
LGTM

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -0,0 +1,886 @@
+  }
+
+  private fun addInteractionForPendingState(
+    pendingItemList: MutableList<StateItemViewModel>, interaction: Interaction, hasPreviousButton: Boolean,
```
These should come in multiple lines for better code clarity:
```
    pendingItemList: MutableList<StateItemViewModel>,
    interaction: Interaction,
    hasPreviousButton: Boolean,
    gcsEntityId: String
```

Ditto elsewhere.

BenHenning replied:
Done, and I cleaned up a bunch of other code here.

rt4914 reviewed (approved):
LGTM. Fix linter issues and it should be good to merge.

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Remove infeasible testing structures, add documentation, and clean up
implementation to prepare for code review.

BenHenning committed:
Add notice that the dispatchers utility is temporary.

BenHenning committed:
Cleanup new LiveData bridge, add tests for it, and migrate other
DataProviders tests to using TestCoroutineDispatchers utility.

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Add AsyncResult tests, fix & re-enable an earlier test in PersistentCacheStoreTest, and fix FakeSystemClock so that it works properly in test environments.

BenHenning committed:
Use ktlint to reformat TestCoroutineDispatchers per reviewer comment
thread.

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Reformat files failing linter check.

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

Conflicts:
	utility/src/main/java/org/oppia/util/data/DataProviders.kt

BenHenning committed:
Reformat new DataProviders code.

BenHenning committed:
Add new tests for the NestedTransformedDataProvider.

BenHenning committed:
Merge branch 'develop' into introduce-new-coroutine-live-data-mechanism

 Conflicts:
	testing/src/main/java/org/oppia/testing/FakeSystemClock.kt

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionAdapter.kt
	utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt

BenHenning committed:
Address reviewer comments & fix lint issue.

BenHenning commented:
Thanks all! @veena14cs sending this back to you to confirm whether I addressed your comments.

Also and also to you @rt4914 for your one comment.

BenHenning committed:
Merge branch 'develop' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'develop' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

rt4914 commented:
Unassigned myself as I have resolved all comments.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into introduce-question-player-ui-structure

veena14cs reviewed (changes requested):
Please assign me once the changes are done as discussed in the comment.

BenHenning committed:
Update hints & solutions handling to more closely follow the intended
behavior graph (and added an actual visual representation of this graph
in code to simplify maintaining the solution in the future). Tests will
be added as part of solving #1273.

BenHenning commented:
@veena14cs & @nikitamarysolomanpvt sending this back to you to resolve your open comment threads. Please let me know if you have any questions.

BenHenning committed:
Merge branch 'develop' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

veena14cs reviewed (approved):
LGTM.
